File: node/src/channel.ts
import type { RPCClient } from './client.js';

export interface ChannelMessage {
  type: 'message' | 'close' | 'error';
  data?: any;
  error?: string;
  sender?: string; // For private channels
}

/**
 * Bidirectional communication channel between RPC clients
 */
export class Channel {
  protected closed = false;
  protected unsubscribe?: () => void;
  protected initialized = false;

  // Return type is `void` (consumer-void) so handlers that incidentally
  // return a value type-check. Runtime uses isPromise() on the actual
  // returned value to detect async handlers and chain them.
  private handlers = new Map<string, Set<(data: any) => void>>();
  private closeHandlers = new Set<() => void>();
  private errorHandlers = new Set<(error: Error) => void>();
  private subscriptions: Promise<() => void>[] = [];

  /**
   * Check if channel is closed
   */
  get isClosed(): boolean {
    return this.closed;
  }

  /**
   * Get the channel ID
   */
  get id(): string {
    return this.channelId;
  }

  constructor(
    protected client: RPCClient,
    protected channelId: string,
  ) {}

  /**
   * Initialize the channel (called by RPCClient)
   */
  public async init(): Promise<void> {
    if (this.initialized) {
      return;
    }

    this.initialized = true;
    const subject = `channel.${this.channelId}`;

    this.unsubscribe = await this.client.subscribe(subject, async (msg: ChannelMessage) => {
      switch (msg.type) {
        case 'message':
          this.emit('message', msg.data);
          break;
        case 'close':
          this.handleClose();
          break;
        case 'error':
          this.handleError(new Error(msg.error ?? 'Channel error'));
          break;
      }
    });
  }

  /**
   * Send data through the channel
   */
  public async send<TMessage = any>(data: TMessage): Promise<void> {
    if (this.closed) {
      throw new Error('Channel is closed');
    }

    const msg: ChannelMessage = {
      type: 'message',
      data,
    };

    await this.client.publish(`channel.${this.channelId}`, msg);
  }

  /**
   * Send a request and wait for reply
   */
  public async request<TRequest = any, TResponse = any>(data: TRequest, timeout = 5000): Promise<TResponse> {
    if (this.closed) {
      throw new Error('Channel is closed');
    }

    const subject = `channel.${this.channelId}.request`;
    return this.client.request<TRequest, TResponse>(subject, data, { timeout });
  }

  /**
   * Setup a request handler for this channel
   */
  public async onRequest<TRequest = any, TResponse = any>(handler: (data: TRequest) => Promise<TResponse> | TResponse): Promise<() => void> {
    const subject = `channel.${this.channelId}.request`;
    const unsubscribe = this.client.onRequest(subject, handler);
    this.subscriptions.push(unsubscribe);
    return unsubscribe;
  }

  /**
   * Listen for messages
   */
  public on<TEvent = any>(event: 'message', handler: (data: TEvent) => void): void;
  public on(event: 'close', handler: () => void): void;
  public on(event: 'error', handler: (error: Error) => void): void;
  public on(event: string, handler: any): void {
    if (event === 'close') {
      this.closeHandlers.add(handler);
    } else if (event === 'error') {
      this.errorHandlers.add(handler);
    } else {
      if (!this.handlers.has(event)) {
        this.handlers.set(event, new Set());
      }
      this.handlers.get(event)!.add(handler);
    }
  }

  /**
   * Remove event listener
   */
  public off<TEvent = any>(event: 'message', handler: (data: TEvent) => void): void;
  public off(event: 'close', handler: () => void): void;
  public off(event: 'error', handler: (error: Error) => void): void;
  public off(event: string, handler: any): void {
    if (event === 'close') {
      this.closeHandlers.delete(handler);
    } else if (event === 'error') {
      this.errorHandlers.delete(handler);
    } else {
      this.handlers.get(event)?.delete(handler);
    }
  }

  /**
   * Close the channel gracefully
   */
  public async close(): Promise<void> {
    if (this.closed) return;

    this.closed = true;

    // Try to notify other side
    try {
      const msg: ChannelMessage = { type: 'close' };
      await this.client.publish(`channel.${this.channelId}`, msg);
    } catch {
      // Ignore publish errors during close
    }

    // Cleanup
    await this.cleanup();
  }

  // Per-channel promise chain. Handlers for message / close / error are
  // serialized via this chain so dispatch stays ordered — matches Python's
  // channel.py:_emit/_handle_close/_handle_error behavior. The handler is
  // invoked INSIDE the chain; invoking eagerly and only chaining the
  // completion would run async handlers concurrently and out of order.
  private handlerChain: Promise<void> = Promise.resolve();

  private runHandler<Arg>(handler: (arg: Arg) => void, arg: Arg, label: string): void {
    this.handlerChain = this.handlerChain.then(() => handler(arg)).catch((error) => console.error(`Error in ${label}:`, error));
  }

  protected emit(event: string, data?: any): void {
    const handlers = this.handlers.get(event);
    if (!handlers) return;
    for (const handler of handlers) {
      this.runHandler(handler, data, 'channel handler');
    }
  }

  protected handleClose(): void {
    if (this.closed) return;

    this.closed = true;

    for (const handler of this.closeHandlers) {
      this.runHandler(() => handler(), undefined, 'close handler');
    }

    this.cleanup().catch(console.error);
  }

  protected handleError(error: Error): void {
    for (const handler of this.errorHandlers) {
      this.runHandler(handler, error, 'error handler');
    }
  }

  protected async cleanup(): Promise<void> {
    // Clear all handlers
    this.handlers.clear();
    this.closeHandlers.clear();
    this.errorHandlers.clear();

    // Clear subscriptions — each element is a Promise resolving to the
    // unsubscribe function; it must be awaited AND invoked.
    await Promise.allSettled(this.subscriptions.map(async (unsub) => (await unsub)()));
    this.subscriptions = [];

    // Unsubscribe from NATS
    if (this.unsubscribe) {
      try {
        this.unsubscribe();
      } catch {
        // Ignore unsubscribe errors
      }
      this.unsubscribe = undefined;
    }

    // Disconnect isolated client if present
    const isolatedClient: RPCClient | undefined = (this as any)._isolatedClient;
    if (isolatedClient) {
      try {
        await isolatedClient.disconnect();
      } catch {
        // Ignore disconnect errors
      }
    }
  }
}

/**
 * Private channel for 1:1 communication between two specific clients
 */
export class PrivateChannel extends Channel {
  private readonly clientId: string;
  private remoteClientId?: string;

  /**
   * Get the remote client ID (if connected)
   */
  get remoteId(): string | undefined {
    return this.remoteClientId;
  }

  constructor(
    client: RPCClient,
    channelId: string,
    private targetClientId: string,
  ) {
    super(client, channelId);
    // Use the original client name, not the isolated connection name
    // When using isolated connections, the name includes suffixes like '-channel-secret-chat'
    // We need to extract the original name to maintain consistent identity
    const fullName = client.options.name ?? `client-${Date.now()}`;
    // Extract base name by removing any suffixes added for isolated connections
    this.clientId = fullName.replace(/-(channel|private)-.*$/, '');
  }

  /**
   * Initialize private channel with handshake
   */
  public async init(): Promise<void> {
    if (this.initialized) {
      return;
    }

    this.initialized = true;

    // Use a unique subject that includes channelId and both client IDs for true privacy
    const sortedIds = [this.clientId, this.targetClientId].sort();
    const subject = `channel.private.${this.channelId}.${sortedIds.join('.')}`;

    this.unsubscribe = await this.client.subscribe(subject, async (msg: ChannelMessage) => {
      // Filter messages: only process if it's for us
      if (msg.sender === this.clientId) {
        // Skip our own messages
        return;
      }

      if (msg.sender !== this.targetClientId) {
        // Only accept messages from the target client
        return;
      }

      if (!this.remoteClientId && msg.sender) {
        // First message establishes the connection
        this.remoteClientId = msg.sender;
      }

      if (this.remoteClientId && msg.sender !== this.remoteClientId) {
        // After connection established, only accept from connected client
        return;
      }

      switch (msg.type) {
        case 'message':
          // Filter out handshake messages
          if (msg.data && typeof msg.data === 'object' && '__handshake' in msg.data) {
            // Handshake received, connection established
            return;
          }
          this.emit('message', msg.data);
          break;
        case 'close':
          this.handleClose();
          break;
        case 'error':
          this.handleError(new Error(msg.error ?? 'Channel error'));
          break;
      }
    });

    // Send initial handshake
    try {
      await this.sendRaw({
        type: 'message',
        data: { __handshake: true },
        sender: this.clientId,
      });
    } catch {
      // Ignore handshake errors - connection might still work
    }
  }

  /**
   * Send data through the private channel
   */
  public async send<TMessage = any>(data: TMessage): Promise<void> {
    if (this.closed) {
      throw new Error('Channel is closed');
    }

    await this.sendRaw({
      type: 'message',
      data,
      sender: this.clientId,
    });
  }

  /**
   * Send a request and wait for reply using native NATS request/reply
   */
  public async request<TRequest = any, TResponse = any>(data: TRequest, timeout = 5000): Promise<TResponse> {
    if (this.closed) {
      throw new Error('Channel is closed');
    }

    const sortedIds = [this.clientId, this.targetClientId].sort();
    const subject = `channel.private.${this.channelId}.${sortedIds.join('.')}.request`;

    return this.client.request(subject, data, { timeout });
  }

  /**
   * Setup a request handler for this private channel
   */
  public async onRequest<TRequest = any, TResponse = any>(handler: (data: TRequest) => Promise<TResponse> | TResponse): Promise<() => void> {
    const sortedIds = [this.clientId, this.targetClientId].sort();
    const subject = `channel.private.${this.channelId}.${sortedIds.join('.')}.request`;

    return this.client.onRequest(subject, handler);
  }

  /**
   * Close the private channel gracefully
   */
  public async close(): Promise<void> {
    if (this.closed) return;

    this.closed = true;

    // Try to notify remote client, but don't fail if it errors
    try {
      await this.sendRaw({
        type: 'close',
        sender: this.clientId,
      });
    } catch {
      // Ignore send errors during close
    }

    await this.cleanup();
  }

  /**
   * Check if channel is connected to a specific client
   */
  public isConnectedTo(clientId: string): boolean {
    return this.remoteClientId === clientId;
  }

  private async sendRaw(msg: ChannelMessage): Promise<void> {
    // Use the same subject format as in init()
    const sortedIds = [this.clientId, this.targetClientId].sort();
    const subject = `channel.private.${this.channelId}.${sortedIds.join('.')}`;
    await this.client.publish(subject, msg);
  }
}

File: go/channel.go
package rpc

import (
	"context"
	"fmt"
	"sort"
	"strings"
	"sync"
	"time"
)

// Channel is a bidirectional broadcast communication channel.
type Channel struct {
	client    *Client
	channelID string
	subject   string

	mu            sync.RWMutex
	closed        bool
	unsub         func()
	requestUnsubs []func()

	onMessage func(data any)
	onClose   func()
	onError   func(err error)

	isolatedClient *Client
}

// NewChannel creates a new broadcast channel. Call Init() before use.
func NewChannel(client *Client, channelID string) *Channel {
	return &Channel{
		client:    client,
		channelID: channelID,
		subject:   "channel." + channelID,
	}
}

// ID returns the channel identifier.
func (ch *Channel) ID() string {
	return ch.channelID
}

// IsClosed returns whether the channel is closed.
func (ch *Channel) IsClosed() bool {
	ch.mu.RLock()
	defer ch.mu.RUnlock()
	return ch.closed
}

// Init subscribes to the channel subject. Must be called before Send/Close.
func (ch *Channel) Init() error {
	unsub, err := ch.client.Subscribe(ch.subject, func(data []byte) {
		var msg ChannelMessage
		if err := DecodeMessageInto(data, &msg); err != nil {
			return
		}

		switch msg.Type {
		case "message":
			if ch.onMessage != nil {
				ch.onMessage(msg.Data)
			}
		case "close":
			ch.handleClose()
		case "error":
			if ch.onError != nil {
				errMsg := msg.Error
				if errMsg == "" {
					errMsg = "Channel error"
				}
				ch.onError(fmt.Errorf("%s", errMsg))
			}
		}
	})
	if err != nil {
		return err
	}
	ch.unsub = unsub
	return nil
}

// Send publishes data to the channel.
func (ch *Channel) Send(data any) error {
	ch.mu.RLock()
	if ch.closed {
		ch.mu.RUnlock()
		return fmt.Errorf("channel is closed")
	}
	ch.mu.RUnlock()

	msg := ChannelMessage{Type: "message", Data: data}
	return ch.client.Publish(ch.subject, msg)
}

// Request sends a request on the channel and waits for a reply.
func (ch *Channel) Request(ctx context.Context, data any, timeout ...time.Duration) ([]byte, error) {
	ch.mu.RLock()
	if ch.closed {
		ch.mu.RUnlock()
		return nil, fmt.Errorf("channel is closed")
	}
	ch.mu.RUnlock()

	t := 5 * time.Second
	if len(timeout) > 0 {
		t = timeout[0]
	}
	return ch.client.Request(ctx, ch.subject+".request", data, t)
}

// OnRequest registers a handler for request/reply on this channel.
// The subscription is tracked so Close()/cleanup releases it.
func (ch *Channel) OnRequest(handler func(data []byte) (any, error)) (func(), error) {
	unsub, err := ch.client.OnRequest(ch.subject+".request", handler)
	if err != nil {
		return nil, err
	}
	ch.mu.Lock()
	ch.requestUnsubs = append(ch.requestUnsubs, unsub)
	ch.mu.Unlock()
	return unsub, nil
}

// OnMessage sets the callback for incoming messages.
func (ch *Channel) OnMessage(fn func(data any)) {
	ch.onMessage = fn
}

// OnClose sets the callback for channel close events.
func (ch *Channel) OnClose(fn func()) {
	ch.onClose = fn
}

// OnError sets the callback for channel errors.
func (ch *Channel) OnError(fn func(err error)) {
	ch.onError = fn
}

// Close gracefully closes the channel.
func (ch *Channel) Close() error {
	ch.mu.Lock()
	if ch.closed {
		ch.mu.Unlock()
		return nil
	}
	ch.closed = true
	ch.mu.Unlock()

	// Notify other side
	msg := ChannelMessage{Type: "close"}
	_ = ch.client.Publish(ch.subject, msg)

	return ch.cleanup()
}

func (ch *Channel) handleClose() {
	ch.mu.Lock()
	if ch.closed {
		ch.mu.Unlock()
		return
	}
	ch.closed = true
	ch.mu.Unlock()

	if ch.onClose != nil {
		ch.onClose()
	}
	_ = ch.cleanup()
}

func (ch *Channel) cleanup() error {
	ch.onMessage = nil
	ch.onClose = nil
	ch.onError = nil

	// Release request/reply subscriptions collected via OnRequest.
	ch.mu.Lock()
	requestUnsubs := ch.requestUnsubs
	ch.requestUnsubs = nil
	ch.mu.Unlock()
	for _, unsub := range requestUnsubs {
		unsub()
	}

	if ch.unsub != nil {
		ch.unsub()
		ch.unsub = nil
	}

	if ch.isolatedClient != nil {
		return ch.isolatedClient.Disconnect()
	}
	return nil
}

// PrivateChannel is a 1:1 communication channel between two specific clients.
type PrivateChannel struct {
	client         *Client
	channelID      string
	clientID       string
	targetClientID string
	subject        string

	mu             sync.RWMutex
	closed         bool
	unsub          func()
	requestUnsubs  []func()
	remoteClientID string

	onMessage func(data any)
	onClose   func()
	onError   func(err error)

	isolatedClient *Client
}

// NewPrivateChannel creates a new private 1:1 channel. Call Init() before use.
func NewPrivateChannel(client *Client, channelID, targetClientID string) *PrivateChannel {
	// Derive clientID from client name, stripping any isolation suffixes
	name := client.Options.Name
	// Remove suffixes like "-channel-xxx" or "-private-xxx"
	for _, suffix := range []string{"-channel-", "-private-"} {
		if idx := strings.Index(name, suffix); idx != -1 {
			name = name[:idx]
		}
	}

	// Sort IDs to create a deterministic subject
	ids := []string{name, targetClientID}
	sort.Strings(ids)
	subject := fmt.Sprintf("channel.private.%s.%s", channelID, strings.Join(ids, "."))

	return &PrivateChannel{
		client:         client,
		channelID:      channelID,
		clientID:       name,
		targetClientID: targetClientID,
		subject:        subject,
	}
}

// ID returns the channel identifier.
func (ch *PrivateChannel) ID() string {
	return ch.channelID
}

// RemoteID returns the remote client ID once connected.
func (ch *PrivateChannel) RemoteID() string {
	ch.mu.RLock()
	defer ch.mu.RUnlock()
	return ch.remoteClientID
}

// IsClosed returns whether the channel is closed.
func (ch *PrivateChannel) IsClosed() bool {
	ch.mu.RLock()
	defer ch.mu.RUnlock()
	return ch.closed
}

// Init subscribes to the private channel and sends a handshake.
func (ch *PrivateChannel) Init() error {
	unsub, err := ch.client.Subscribe(ch.subject, func(data []byte) {
		var msg ChannelMessage
		if err := DecodeMessageInto(data, &msg); err != nil {
			return
		}

		// Filter: skip own messages
		if msg.Sender == ch.clientID {
			return
		}
		// Only accept from target
		if msg.Sender != ch.targetClientID {
			return
		}

		ch.mu.Lock()
		if ch.remoteClientID == "" && msg.Sender != "" {
			ch.remoteClientID = msg.Sender
		}
		remote := ch.remoteClientID
		ch.mu.Unlock()

		if remote != "" && msg.Sender != remote {
			return
		}

		switch msg.Type {
		case "message":
			// Filter handshake messages
			if m, ok := msg.Data.(map[string]any); ok {
				if _, isHandshake := m["__handshake"]; isHandshake {
					return
				}
			}
			if ch.onMessage != nil {
				ch.onMessage(msg.Data)
			}
		case "close":
			ch.handleClose()
		case "error":
			if ch.onError != nil {
				errMsg := msg.Error
				if errMsg == "" {
					errMsg = "Channel error"
				}
				ch.onError(fmt.Errorf("%s", errMsg))
			}
		}
	})
	if err != nil {
		return err
	}
	ch.unsub = unsub

	// Send handshake
	_ = ch.sendRaw(ChannelMessage{
		Type:   "message",
		Data:   HandshakeData{Handshake: true},
		Sender: ch.clientID,
	})

	return nil
}

// Send publishes data to the private channel.
func (ch *PrivateChannel) Send(data any) error {
	ch.mu.RLock()
	if ch.closed {
		ch.mu.RUnlock()
		return fmt.Errorf("channel is closed")
	}
	ch.mu.RUnlock()

	return ch.sendRaw(ChannelMessage{
		Type:   "message",
		Data:   data,
		Sender: ch.clientID,
	})
}

// Request sends a request on the private channel and waits for a reply.
func (ch *PrivateChannel) Request(ctx context.Context, data any, timeout ...time.Duration) ([]byte, error) {
	ch.mu.RLock()
	if ch.closed {
		ch.mu.RUnlock()
		return nil, fmt.Errorf("channel is closed")
	}
	ch.mu.RUnlock()

	t := 5 * time.Second
	if len(timeout) > 0 {
		t = timeout[0]
	}

	ids := []string{ch.clientID, ch.targetClientID}
	sort.Strings(ids)
	subject := fmt.Sprintf("channel.private.%s.%s.request", ch.channelID, strings.Join(ids, "."))
	return ch.client.Request(ctx, subject, data, t)
}

// OnRequest registers a handler for request/reply on this private channel.
// The subscription is tracked so Close()/cleanup releases it.
func (ch *PrivateChannel) OnRequest(handler func(data []byte) (any, error)) (func(), error) {
	ids := []string{ch.clientID, ch.targetClientID}
	sort.Strings(ids)
	subject := fmt.Sprintf("channel.private.%s.%s.request", ch.channelID, strings.Join(ids, "."))
	unsub, err := ch.client.OnRequest(subject, handler)
	if err != nil {
		return nil, err
	}
	ch.mu.Lock()
	ch.requestUnsubs = append(ch.requestUnsubs, unsub)
	ch.mu.Unlock()
	return unsub, nil
}

// OnMessage sets the callback for incoming messages.
func (ch *PrivateChannel) OnMessage(fn func(data any)) {
	ch.onMessage = fn
}

// OnClose sets the callback for channel close events.
func (ch *PrivateChannel) OnClose(fn func()) {
	ch.onClose = fn
}

// OnError sets the callback for channel errors.
func (ch *PrivateChannel) OnError(fn func(err error)) {
	ch.onError = fn
}

// Close gracefully closes the private channel.
func (ch *PrivateChannel) Close() error {
	ch.mu.Lock()
	if ch.closed {
		ch.mu.Unlock()
		return nil
	}
	ch.closed = true
	ch.mu.Unlock()

	_ = ch.sendRaw(ChannelMessage{
		Type:   "close",
		Sender: ch.clientID,
	})

	return ch.cleanup()
}

// IsConnectedTo checks if the channel is connected to a specific client.
func (ch *PrivateChannel) IsConnectedTo(clientID string) bool {
	ch.mu.RLock()
	defer ch.mu.RUnlock()
	return ch.remoteClientID == clientID
}

func (ch *PrivateChannel) handleClose() {
	ch.mu.Lock()
	if ch.closed {
		ch.mu.Unlock()
		return
	}
	ch.closed = true
	ch.mu.Unlock()

	if ch.onClose != nil {
		ch.onClose()
	}
	_ = ch.cleanup()
}

func (ch *PrivateChannel) cleanup() error {
	ch.onMessage = nil
	ch.onClose = nil
	ch.onError = nil

	// Release request/reply subscriptions collected via OnRequest.
	ch.mu.Lock()
	requestUnsubs := ch.requestUnsubs
	ch.requestUnsubs = nil
	ch.mu.Unlock()
	for _, unsub := range requestUnsubs {
		unsub()
	}

	if ch.unsub != nil {
		ch.unsub()
		ch.unsub = nil
	}

	if ch.isolatedClient != nil {
		return ch.isolatedClient.Disconnect()
	}
	return nil
}

func (ch *PrivateChannel) sendRaw(msg ChannelMessage) error {
	return ch.client.Publish(ch.subject, msg)
}

// Channel creates or joins a broadcast channel.
func (c *Client) Channel(channelID string, isolated ...bool) (*Channel, error) {
	client := c
	var isolatedClient *Client

	if len(isolated) > 0 && isolated[0] {
		client = c.createIsolatedClient("channel-" + channelID)
		c.isolatedClientsMu.Lock()
		c.isolatedClients = append(c.isolatedClients, client)
		c.isolatedClientsMu.Unlock()
		if err := client.Connect(context.Background()); err != nil {
			return nil, err
		}
		isolatedClient = client
	}

	ch := NewChannel(client, channelID)
	ch.isolatedClient = isolatedClient
	if err := ch.Init(); err != nil {
		if isolatedClient != nil {
			_ = isolatedClient.Disconnect()
			c.removeIsolatedClient(isolatedClient)
		}
		return nil, err
	}
	return ch, nil
}

// PrivateChannelConnect creates or joins a private 1:1 channel.
func (c *Client) PrivateChannelConnect(channelID, targetClientID string, isolated ...bool) (*PrivateChannel, error) {
	client := c
	var isolatedClient *Client

	if len(isolated) > 0 && isolated[0] {
		client = c.createIsolatedClient("private-" + channelID)
		c.isolatedClientsMu.Lock()
		c.isolatedClients = append(c.isolatedClients, client)
		c.isolatedClientsMu.Unlock()
		if err := client.Connect(context.Background()); err != nil {
			return nil, err
		}
		isolatedClient = client
	}

	ch := NewPrivateChannel(client, channelID, targetClientID)
	ch.isolatedClient = isolatedClient
	if err := ch.Init(); err != nil {
		if isolatedClient != nil {
			_ = isolatedClient.Disconnect()
			c.removeIsolatedClient(isolatedClient)
		}
		return nil, err
	}
	return ch, nil
}
